### ISSUE TITLE
Multiple Images

### ISSUE THREAD

**Issue by riker09** [TandoorRecipes/recipes#1446]
### Describe the solution you'd like

I'd like to have more than one image per recipe. On the recipe detail view there could be a small slideshow of all pictures (if more than one is availalbe, of course)

### Contribute

- [ ] Yes
- [X] Partly
- [ ] No

**Comment by vabene1111** [comment]
you can add images to steps so, is that sufficent for you ?

**Comment by JSAnyone** [comment]
I think having multiple images to swipe between would be really nice. 
I don't really see the point in having images for steps, what's the use case for this?

**Comment by smilerz** [comment]
>  I don't really see the point in having images for steps, what's the use case for this?

A complicated step that requires a lot of manual manipulation of ingredients is the case I would use it for most commonly.  Making Asian dumplings as one example.

**Comment by riker09** [comment]
Hello. I just tried to verify that on my Android phone. I'm not able to upload an image to a step. Are you referring to the textarea wysiwyg editor? I have not tried that.

This aside, I don't think that uploading images to steps would solve my use case. I simply would like to be able to add more than one picture to a recipe. On the recipe detail page there could be carousel that infity-scrolls through all available images. On the overview page a second picture could be shown when hovering an item with the mouse. I don't know a good way to present that on a mobile device, but I bet there is a way.

I believe one goal for me would be to see my process as a cook. This kinda touches the "protocol" feature where I can rate my cooking. Maybe this dialog should have the option to upload a picture as well?

**Comment by vabene1111** [comment]
uploading images to steps schuld work trough the "file" property of any given step but i have not tested that on android yet.

Multiple images for recipes are fine (altough i personally would not use them as steps) but shoudl be something we can migrate to in the future. We will generally need to keep an eye on how files are used in conjunction with recipes (i personally prefer to have the least amount of files possible but i also want to offer flexebility to those who want to use them)

**Comment by dlaukamp** [comment]
I installed my tandoor instance only a couple days ago and starting migrating my existing recipes into it. While I am really glad that tandoor makes this process so smooth in many details and was pleasantly surprised on how easy it is to reuse ingredients/ units/ etc., I searched quite a while for the option to add multiple images to the recipe, thinking that I must have missed such an obvious feature.

Apparently, it doesn't seem to be that obvious or even in-demand, given that this issue is only a couple days old and does not seem to be a duplicate of similar requests in the past. Still, isn't it natural to expect that you could add a picture after re-cooking an existing recipe, showcasing your success (or at least genuine attempt) in reproducing the result?

> And yes, this is very much no high-prio issue. I just thought it would make a good addition.

Basically this, definitely not a deal breaker.

**Comment by B5m3M** [comment]
I'm looking forward to a multiple-image feature. 
I love taking several Photos of the meals I cooked. For many not needed, for me it was almost a deal breaker if it wasn't for so many other great features in Tandoor.

**Comment by smilerz** [comment]
@dlaukamp & @B5m3M - given the use case you provided, would you like to have the pictures as part of the cooklog?

**Comment by vabene1111** [comment]
@smilerz i think we should just migrate everything to the `UserFiles` system. This way we could add relations to any object to allow it to have files. 

I must admit the main reason not doing this before is that i am afraid of breaking things with files, as it is much harder to make migrations there than when having Database migrations and while many people backup their database `mediafiles` are often neglected. So making this change requires a very very very high degree of caution and testing and thinking of all possible edge cases

**Comment by dlaukamp** [comment]
> @dlaukamp & @B5m3M - given the use case you provided, would you like to have the pictures as part of the cooklog?

At least for me, I think of these pictures as a representation of the recipe itself, not as a documentation of when I did cook a specific recipe. I just created some entries in the cooklog for testing purposes and find it hard to imagine how to integrate the images there, purely from a layout perspective.

I think the original poster did already nail it - if multiple images are available, have a slideshow instead of a static picture right at the top.

> the main reason not doing this before is that i am afraid of breaking things with files, as it is much harder to make migrations there than when having Database migrations

Thank you for the explanation, I couldn't really make sense of the earlier "I prefer to have the least amount of files".

**Comment by rubenixnagios** [comment]
> you can add images to steps so, is that sufficent for you ?

Im not able to add images to steps...anyone can send further instructions?
@vabene1111 is that a link somewhere to add images? can't find it...
thanks in advance!

**Comment by rubenixnagios** [comment]
thanks @smilerz and sorry for posting before updating to latest release!!
once updated can see  File Button!! lovely!! great work!! I translated to catalan a long time ago so there will be loads of strings untranslated...

**Comment by vabene1111** [comment]
thanks for your contribution, you can always hop on translate.tandoor.dev and translate some more if you want or just enjoy tandoor in english/mixed.

**Comment by rubenixnagios** [comment]
> thanks for your contribution, you can always hop on translate.tandoor.dev and translate some more if you want or just enjoy tandoor in english/mixed.

alright, I will do a little by a little as am quite busy adding family recipes to tandoor  xDDD
catanglish tandoor does not look that bad anyway ;P

**Comment by HarHarLinks** [comment]
I would like to add multiple images/videos/files to some steps, and have them look nice like a gallery.

**Comment by rubenixnagios** [comment]
already posted some new strings to translate.tandoor.dev
will this be available on next release? do i have to create a pull request?
thanks again!! great job

**Comment by vabene1111** [comment]
If you added something on the translate system it should automatically be included in the next release.

**Comment by Maschine2** [comment]
I also support the request for multiple images shown as a slideshow or "gallery" on the recipe level (not hidden in steps). My use case is to show different views of the same cooked recipe or different varieties from the same recipe:

- If I bake a bread, I want to show the full (baked) bread, the bread cut in half (to show the inside), and maybe a third picture with a nicely garnished slice of bread with butter, cheese, tomato, etc.
- For a "basic" recipe like a Pizza dough, you could show different varieties of Pizza like Pizza Margherita, Cheese Pizza, or Saussage Pizza.

**Comment by vabene1111** [comment]
understood. I personally think that I want to move recipe images to the "UserFile" system so that I have one generic way of handling all files, compression and so on.

**Comment by igorsantos07** [comment]
I don't feel like the "file" option under steps is an ideal replacement for "step pictures", as it is right now on the editor. Suggestions:
- make the name optional - trying to upload without one simply shows a generic error message
- show the name under the pic, as a legend of sorts
- allow adding more than one file per step - which _feels_ it should be possible, given there's a `[+]` button besides the selectbox, but it's not possible. Actually, that selectbox is confusing, as it doesn't show any option? If the idea is to show a system-wide list of files... it doesn't feel right with pic upload either.
- allow removal of the current file
- name that button "File / image", so it's more obvious when someone is looking for that
- add a note below the image field in the recipe editor, mentioning further pictures can be added in each step

**Comment by hujuDev** [comment]
Is the "file" option for steps working as intended? When I try to use it (in order to add per-step images), it just shows an empty dropdown with no way for me to add a file or an image. 

<img width="1460" height="262" alt="Image" src="https://github.com/user-attachments/assets/aae0a566-6cbd-449e-8112-238c451ef198" />

Ideally I'd love it if the per-step images for a recipe are automatically imported from a URL, just like the title image, is something like that on the horizon?

**Comment by vabene1111** [comment]
the UI is just bad right now, you need to upload a custom file first trough the database > files view and then you can link it to a setp

**Comment by despaws** [comment]
I also support the request for multiple images for each recipe.  I'm using other recipe apps right now that support it, and one of the primary uses I have for it is we've got hundreds of 3x5 recipe cards handed down through extended friends and family over the years that I've used Gemini to scan and convert to recipes and imported to these other recipe apps.  The ability to have multiple images allows me to have a picture of the dish as the "primary" photo, but then also have pictures of the front and back of the old 3x5 card for reference (and posterity, nostalgia, etc).

I'm playing with Tandoor right now and like what I see, but the inability to bring over those historical 3x5 cards as extra images on the recipe is a big drawback.

**Comment by smilerz** [comment]
> I also support the request for multiple images for each recipe. I'm using other recipe apps right now that support it, and one of the primary uses I have for it is we've got hundreds of 3x5 recipe cards handed down through extended friends and family over the years that I've used Gemini to scan and convert to recipes and imported to these other recipe apps. The ability to have multiple images allows me to have a picture of the dish as the "primary" photo, but then also have pictures of the front and back of the old 3x5 card for reference (and posterity, nostalgia, etc).
> 
> I'm playing with Tandoor right now and like what I see, but the inability to bring over those historical 3x5 cards as extra images on the recipe is a big drawback.

you can add them to a step - just not the headline photo
